ENH: Extend Rotation.align_vectors() to allow an infinite weight, as well as minimum-distance rotations for single vectors

scottshambaugh opened the pull request:
#### Reference issue
Closes https://github.com/scipy/scipy/issues/17462
Closes https://github.com/scipy/scipy/issues/16880

#### What does this implement/fix?
This implements [the "align-constrain" algorithm](https://digitalcommons.usu.edu/cgi/viewcontent.cgi?article=4112&context=smallsat) for defining a rotation. The user passes in primary and secondary vectors in reference frames A and B. The algorithm calculates a rotation from B to A that exactly aligns the primary vectors, and then rotates about that axis to most closely align the constrained secondary vectors.

This algorithm is equivalent to `align_vectors()` with sets of two vectors, an infinite weight on the first pair of vectors, and a finite nonzero weight on the second pair of vectors. There is a test included here to verify that internal consistency.

#### Additional information
~~I would appreciate some feedback on how to handle the case where primary and secondary vectors are parallel. Currently I have it defined such that if the secondary vector is parallel to the primary to within an angular tolerance `atol`, the rotation is unconstrained and so the vector [1, 0, 0] is arbitrarily chosen as the secondary (followed by [0, 1, 0] if that is also parallel). Another option would be to error out if the vectors are parallel. A third option would be to error out if the user passes in `atol = None` and choose the secondary axis otherwise.~~

Update: The functionality is modified so that in the instance of parallel pairs, only the primary vectors are aligned via a rotation that has no change in angle about the primary axis.

scottshambaugh commented:
This is good to go.

scottshambaugh commented:
This is good to go.

scottshambaugh commented:
To work better with the idea in https://github.com/scipy/scipy/issues/16880, here are the outstanding TODOs:
* Allow for single vector inputs for an unconstrained secondary axis
* Use that code as a test case

scottshambaugh commented:
I've done a final pass over the implementation, tests, and docs, and think this is ready for merge.

tylerjereddy commented:
The branching for the next release is imminent, and there aren't even code reviews here yet, so I'm going to have to bump the milestone.

While I did try to manually review a bunch of PRs last week, obviously bandwidth is finite and I have to draw the line somewhere for releases to actually happen, so this isn't an opinion on the quality of the code. You may also try pinging @nmayorov as I noted in another PR of yours.

nmayorov commented:
Here is my take on that.

What if we generalize and embed this method into the existing `align_vectors` method?

Such that `weights` in `align_vectors` are allowed to have a _single_ element with infinity, meaning that the corresponding pair of vectors must be aligned exactly. And the remaining vectors must be aligned as best as possible according to the same quadratic cost function. No restriction on the total number of vectors, the algorithm with 2 vectors is a special case. 

Looks like the problem is well defined in this form and we should be able to find the algorithm to solve it. The `atol` parameter probably will be gone, but we can think about it later.

@scottshambaugh what is your preliminary opinion on that?

scottshambaugh commented:
I quite like that as a special case of `align_vectors`, and would be happy to add that. As currently written, an initial weight of infinity would essentially be the same as the call `align_constrain([pri0, sec0], align_vectors([[pri1, sec1], .....], )`, and that's easy enough to put in. 

I do think it's best to keep the `align_constrain` method separate here however. I don't think an initial weight of infinity is an intuitive way to think about the problem even if it is equivalent, and in my experience with attitude determination you would reach for these methods for very different purposes. `align_vectors` to combine a large number of attitude estimates into a single best estimate, and `align_constrain` to figure out idealized rotations to point your reference vectors at their targets. 

The `atol` will need to remain though, since it is protecting for floating point inaccuracies around a divide by zero. Otherwise, I would advocate for splitting the case where you align only primary vectors via a minimal angle rotation into its own method (but perhaps that is a good idea too).

nmayorov commented:
> be the same as the call align_constrain([pri0, sec0], align_vectors([[pri1, sec1], .....], ), and that's easy enough to put in.

Not sure if I understand. `align_vectors` returns a rotation matrix (besides other things). And your method accepts a set of vectors. I guess you put some conceptual meaning to that, which I don't get immediately.

> I do think it's best to keep the align_constrain method separate here however. I don't think an initial weight of infinity is an intuitive way to think about the problem even if it is equivalent, and in my experience with attitude determination you would reach for these methods for very different purposes. align_vectors to combine a large number of attitude estimates into a single best estimate, and align_constrain to figure out idealized rotations to point your reference vectors at their targets. More modular and easier to think about separately IMO.

I see your point, but embedding infinite weights into the existing algorithm has other benefits:

- No API spread, the method remains more visible, everything in one place, no duplication
- This generalization with infinity is hackish (not in implementation though), but I think it goes well with applied computation packages like scipy (it's no pure math). Someone will find it's even more intuitive I guess
- The API will be naturally generalized (arbitrary number of vectors, one pair of vectors might be "strict")

> The atol will need to remain though, since it is protecting for floating point inaccuracies around a divide by zero (at both the lines _normalize3(w) and rot_vec = cross / np.linalg.norm(cross) * theta). Otherwise, I would advocate for splitting the case where you align only primary vectors via a minimal angle rotation into a third method. But perhaps that is a good idea too - align_vector?

I think everything can be implemented without unstable computations, however it's true that it can be a case where the rotation can be poorly defined (similar to computing atan2(y, x) where x and y are close to zero).

I don't think that 3 methods are good idea, I'm more of an advocate of few, but powerful, methods. 

Aligning a single pair of vectors already works properly in `align_vectors`, doesn't it?

----

I've written a short note on how I see the implementation of constraint alignment with arbitrary number of secondary vectors: https://github.com/scipy/scipy/files/12191058/constrained_alignment.pdf 

I believe the computation of the initial alignment can be made "singularity free", I've used some normalization there, which in general is not good.

Perhaps this can communicate my idea better (or maybe you already have the same in mind)

Also I know a related paper "Pointing in Real Euclidean Space", check it too.

scottshambaugh commented:
Thanks for the links to your doc and the "[Pointing in Real Euclidean Space](https://www.researchgate.net/profile/Daniel-Hershkowitz/publication/245432853_Pointing_in_Real_Euclidean_Space/links/0046351ee2e34c2f82000000/Pointing-in-Real-Euclidean-Space.pdf)" paper, I read through both and they seem good and correct and what we want to be shooting for here. 

I've added [a commit](https://github.com/scipy/scipy/pull/17542/commits/784c7fbbcafff00459b846caf134aa957214a21e) that shows how I see the different functionalities and how they relate / are subsets of each other. I can combine these all into the `align_vectors` function with specific subconditions if that's how the maintainers prefer, right now they are separate just for illustration.

While a single pair of vectors does give a result in `align_vectors`, it does _not_ return a minimal rotation like #16880 asks for. I have some example code here: https://github.com/scipy/scipy/issues/16880#issuecomment-1503770562. This is because aligning a single pair of vectors is underconstrained since you could have any secondary rotation angle about those vectors. Section IV of that paper talks about this:
![image](https://github.com/scipy/scipy/assets/14363975/1a427b5d-2b6b-4f37-80b2-c38f3eb3c576)

So, dividing by the norm of the cross product is needed for that minimal angle calculation, and that cross product norm being zero does represent a real physical singularity. The identity rotation is the shortest angle rotation in that limit, so it makes sense to return that if we get close to the singularity. It might make sense to not make `atol` a parameter and instead calculate it based on the floating point accuracy of the machine running the code, but I don't think we can get rid of `atol` altogether.

nmayorov commented:
> I've added a commit that adds the infinity weight, and shows how I see the different functions relating to / being subsets of each other. I can combine these 3 functions all into the align_vectors function with specific subconditions if that's how the maintainers prefer - right now they are separate just for illustration.

My personal preference is to keep the related algorithms in one method here. I think `align_vectors` is barely making it into API, i.e. it wouldn't be a mistake or bad decision to not even include it from the start. Adding new ways to use it would "strengthen" it, which I think is useful and appropriate in this case.

> While a single pair of vectors does give a result in align_vectors, it does not return a minimal rotation like https://github.com/scipy/scipy/issues/16880 asks for. I have some example code here that shows this: https://github.com/scipy/scipy/issues/16880#issuecomment-1503770562. This is because aligning a single pair of vectors is underconstrained since you could have any secondary rotation angle about those vectors. Section IV of that paper talks about this:

I think it's acceptable to change the behavior specifically for a single vector pair to compute the smallest rotation. But do you think it's important or useful enough to chose this smallest rotation?

> So, dividing by the norm of the cross product is needed for that minimal angle calculation, and that cross product norm being zero does represent a real physical singularity. The identity rotation is the shortest angle rotation in that limit, so I think it makes sense to return that if we get close to the singularity. It might make sense to not make atol a parameter and instead calculate it based on the floating point accuracy of the machine running the code, but I don't think we can get rid of atol altogether.

I thought about it a little bit. 

1. If either `a` and `b` is zero, then we can consider that the constraint is satisfied. However I can't think of a proper way to avoid comparison with a threshold value here. We can use a condition that norms of `a` and `b` is higher that some `eps` and then either raise an error or warn and proceed without satisfying `a` and `b` alignment.
2. If `a` and `b` are not zero, then we can compute their directions by normalizing. Then `c = a X b` is directed along the rotation axis and has the norm `sin(theta)`. Then we compute `theta = atan2(sin(theta), cos(theta))` . To convert `c` to the rotation vector we need to multiply it by `theta / sin(theta)`. This expression can be either computed directly for large enough `theta` or by Taylor expansion as `1 + theta**2/6 + ...`. This way we can always compute the correct rotation vector even for parallel or almost parallel `a` and `b`.

So we need a threshold to cutoff super small vectors, but parallelism can be handled nicely with Taylor expansion without introducing hard thresholds.

I don't see that passing extremely small vectors into the function is an important case and a default threshold with a warning is probably sufficient. But if you see it differently, introducing an extra `atol` named argument is not big a problem.

scottshambaugh commented:
@nmayorov I've combined all the functionality into the `align_vectors` method, and added an example for the baseline behavior there. Now that it's written I agree with your thought that it works best combined into a single powerful method - I think that it flows quite nicely and should hopefully be a bit more useful now.

I switched over to a Taylor series approximation for the small angle in the one instance (good idea!), but think that we should keep the `atol` parameter otherwise. I can think of use cases that use microarcseconds (5e-12 radians), or deliberately align the primary and secondary vector (0 radians), and think it's cleanest to just give the user control rather than picking an arbitrary threshold to raise warnings.

> I think it's acceptable to change the behavior specifically for a single vector pair to compute the smallest rotation. But do you think it's important or useful enough to chose this smallest rotation?

Yes, I think it's useful behavior, and the most intuitive result. Doing anything else would be assuming additional secondary axis constraints / arbitrarily choosing a result from the infinitude of possible rotations.

I think the PR is ready for final review if you agree with the decisions here!

scottshambaugh commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1258,6 +1267,7 @@ def test_align_vectors_noise():
     n_vectors = 100
     rot = Rotation.random(random_state=rnd)
     vectors = rnd.normal(size=(n_vectors, 3))
+    vectors /= np.linalg.norm(vectors, axis=1).reshape(n_vectors, 1)
```
@nmayorov I added a normalization step in `align_vectors` because I believe the calculation of `rssd` should not be dependent on the vector length (if a set of vectors of varying length can align exactly, I would expect rssd to be 0). Let me know if I'm wrong on that

nmayorov replied:
As far as I remember, originally it was done in this method (with a flag whether vectors are already normalized).

However after some discussion this logic was removed and I think it's a correct decision. The reason is that we don't need to have normalized vectors to have a well defined problem. Yes, if vectors are not normalized even with perfect (angular) alignment the cost function is not zero. However this is a transparent behavior which can be completely avoided by normalizing the vectors beforehand. The decision of whether or not normalize vectors are up to the user.

At this point, I think, it's certainly a very undesirable to change this behavior. 

It doesn't seem to create any logical issues for the extensions you are working on, does it?

scottshambaugh replied:
I don't quite understand the use case where you want rssd weighted by vector length, but I agree that it's more general. I removed the normalization step and added some information to the docstring for `rssd` to make the behavior clear.

nmayorov replied:
Originally I though the same. But if there is a need to align vectors of different lengths (i.e. implicit weighing), why not provide this option. And then it's better not to complicate things by adjusting `rssd` definition.

scottshambaugh commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1240,16 +1249,16 @@ def test_align_vectors_improper_rotation():
+    est2, rssd2, cov2 = Rotation.align_vectors(a, b, scale * np.ones(n), True)
 
     assert_allclose(est1.as_matrix(), est2.as_matrix())
-    assert_allclose(np.sqrt(2) * rmsd1, rmsd2)
```
Because the original test formed `a` via an exact rotation of `b`, rmsd was always 0, so this was an ineffective assert. Updated to better test this.

nmayorov commented on scipy/spatial/transform/_rotation.pyx:
```diff
@@ -2766,33 +2806,65 @@ cdef class Rotation:
+        There are two special cases. The first is when one of the weights is
+        infinite. In this case, the rotation is estimated by first applying the
+        Kabsch algorithm to the subset of vectors with finite weights. The
+        rotation is then found from that result, to the infinite-weight vectors
```
What does it mean exactly "rotation is then found from that result"?

Does it correspond to the same optimization problem with the constraint?

scottshambaugh replied:
I reworded this because yeah it was a bit confusing. Basically, the baseline Kabsch algorithm is run on all the non-infinite weights, and then the resulting vectors from that rotation matrix are taken as the single pair of non-infinite vectors for the align-constrain case. I added some tests in `test_align_vectors_near_inf` to confirm approximate equality of a set of 3 vectors with `weights=[1e6, 2, 1]` and `weights=[np.inf, 2, 1]`

nmayorov commented on scipy/spatial/transform/_rotation.pyx:
```diff
@@ -2766,33 +2806,65 @@ cdef class Rotation:
+        The second special case is if a single vector is given for `a` and `b`,
+        in which the shortest distance rotation that aligns `b` to `a` is
+        returned. Any input for `weights` is ignored in this case. This is also
+        the behavior if `a[0]` and `a[1]` or `b[0]` and `b[1]` as found in the
```
Why do you think it's beneficial to consider two vectors as a special case for "constrain" algorithm, especially with tolerance? What if we get 10 almost parallel vectors? 

From the implementation perspective I don't like the fact that we need to implement the whole another function to support it, whereas it should be handled correctly with a general algorithm as far as I understand.

nmayorov replied:
Now I've read the code, I understand what is your approach here.

Now need to understand why it works.

nmayorov commented on scipy/spatial/transform/_rotation.pyx:
```diff
@@ -2766,33 +2806,65 @@ cdef class Rotation:
         Parameters
         ----------
-        a : array_like, shape (N, 3)
+        a : array_like, shape (3,) or (N, 3)
```
Accepting a 1D vector seems fine.

nmayorov commented on scipy/spatial/transform/_rotation.pyx:
```diff
@@ -2766,33 +2806,65 @@ cdef class Rotation:
         weights : array_like shape (N,), optional
             Weights describing the relative importance of the vector
-            observations. If None (default), then all values in `weights` are
+            observations. If `None` (default), then all values in `weights` are
```
I think None typically written as it is, without single or double quotes.

nmayorov commented on scipy/spatial/transform/_rotation.pyx:
```diff
@@ -2806,54 +2878,252 @@ cdef class Rotation:
                                 a.shape, b.shape))
 
+        # Find number of vector pairs, and normalize to unit length
+        if a.shape == (3,):
```
An alternative way to handle the cases is to convert the array to 2D form from the start using `np.atleast_2d`. Maybe it will be cleaner.

scottshambaugh replied:
I agree, that's cleaner

nmayorov commented on scipy/spatial/transform/_rotation.pyx:
```diff
@@ -2806,54 +2878,252 @@ cdef class Rotation:
+            b3 = b
+        else:
+            N = len(a)
+            a /= np.linalg.norm(a, axis=1).reshape(N, 1)
```
I've commented on normalization in other place.

nmayorov commented on scipy/spatial/transform/_rotation.pyx:
```diff
@@ -2806,54 +2878,252 @@ cdef class Rotation:
+                R, _ = cls.align_vectors(a[~weight_inf], b[~weight_inf],
+                                         weights=weights[~weight_inf])
+                R_mat = R.as_matrix()
+                a_non_inf = R_mat[0, :]  # Already normalized
```
Hmm, I can see why it can work, but do you have any mathematical details why it corresponds to the original alignment problem with constraints?

scottshambaugh replied:
I'm not quite sure this is working correctly... need to think about it.

scottshambaugh replied:
Yes this is definitely incorrect right now, figuring out how to fix it.

nmayorov replied:
My sort of intuitive issues with that was: why take axis 0 from `R_mat` here, why not axis 1 or 2?

nmayorov commented on scipy/spatial/transform/_rotation.pyx:
```diff
@@ -2806,54 +2878,252 @@ cdef class Rotation:
+                rssd = np.sqrt(np.sum(weights_non_inf * (a_est - a)**2))
+
+                if return_sensitivity:
+                    # Note: the sensitivity is not meaningful with an infinite
```
I don't think it's "meaningful" per se. But I don't quite understand how to compute it here.

Looks like there should be 2 sensitivities: for the "strict" vectors and for other vectors. Like "how the rotation matrix changes, if we perturb the strict vectors" and "how the rotation matrix changes, if perturb other vectors".

nmayorov replied:
As for the handling. What if we raise an error when `sensitivity=True` and an infinite weight presents?

Could it be better than returning infinities? I'm not sure, think about it.

scottshambaugh replied:
My goal with the behavior of the code path with an infinite weight, is that it would return the same result as `weights = [w0, w1, w2, ...]` in the limit of `w0 -> inf`. With the existing algorithm, I see that as I increase one of the weights towards infinity, the entries in the sensitivity matrix also approach infinity. So I think this accurately captures the behavior in the limit. We could also raise an error, what do you think?

I think your comment about the sensitivities for the strict vs other vectors is a good idea, but I'm not sure how to rigorously think about it.

nmayorov replied:
I think it's possible to compute these 2 sensitivities, but I don't like the idea of changing the number of returned objects once more. (I think it was a bad pattern to choose here in the first place).

I would probably start with an exception for this case, because we don't know what to return.

scottshambaugh replied:
Sounds good to me, changed this to throw an exception.

nmayorov commented on scipy/spatial/transform/_rotation.pyx:
```diff
@@ -2806,54 +2878,252 @@ cdef class Rotation:
+        # If we have only one vector pair, we exactly align `b` to `a` using
+        # the shortest path rotation
+        if N == 1:
+            theta = acos(np.dot(a3, b3))
```
I don't understand one thing.

The constrained algorithm with 2 vectors does something very similar on the first step (aligning primary vectors). But here it is another code branch and `_matrix_from_pri_sec_vectors` works separately and differently. 

Why?

scottshambaugh replied:
The first path with two vectors only computes this angle between `a[0]` and `[a1]`, and `b[0]` and `[b1]`, for the purpose of checking for parallelism. `_matrix_from_pri_sec_vectors` is then used to calculate the A and B frames to impose that secondary axis constraint.

We could use a similar code path for the 1-D case, by doing something like the below:
```
A = np.array(_matrix_from_pri_sec_vectors(a[0], [1, 0, 0]))
B = np.array(_matrix_from_pri_sec_vectors(b[0], [1, 0, 0]))
R_opt = cls.from_matrix(A.T @ B)
```

But then the `[1, 0, 0]` secondary axis is arbitrary, runs into potential issues if it is parallel to one of the other vectors, and will generally not correspond to the minimal angle rotation.

nmayorov replied:
I've outlined the 2 steps algorithm for 1 strict + n other alignment: first compute the strict rotation (as in this code path), then align other vectors (if there are any) using essentially 2D alignment (with atan2). 

Not necessary to follow this path, but considering there are apparent issues with reusing standard alignment on step 2 (discussed above), maybe it's a way to go.

nmayorov commented:
@scottshambaugh 

What's your thoughts on the current status? What out plans would be?

scottshambaugh commented:
Hi @nmayorov, I'm having trouble getting the results between methods to line up, in the case where we have the 1 strict primary and n > 2 secondary alignment.

```python
# Test case unit vectors
a = [[-0.24110296,  0.40209011, -0.88328529], [ 0.93920879, -0.34292415, -0.01702585], [0.71251997, 0.04947096, 0.69990565]]
b = [[ 0.11038608, -0.80648698,  0.58085597], [0.97015365, 0.11062565, 0.21578661], [ 0.4294927 , -0.75374513, -0.49739752]]
weights = [np.inf, 2, 1]
```

My three methods for calculating the above are
1) Setting `weights[0] = 1e12` and running the standard align_vectors algorithm.
2) Find the average of the weighted non-inf a and b vectors, eg `a_non_inf = np.mean(a_non_inf * weights_non_inf.T, axis=0)`. Then calculate the constrained rotation as in the N=2 case.
3) For each of the pairs `(a[0], a[n])` and `(b[0], b[n])`, calculate the constrained rotation as in the N=2 case. Then concatenate them into a single Rotation object R and take the weighted average `R.mean(weights=weights_non_inf)`.

I would expect these to give the same result to some numerical precision, and if I set the third weight to something small like `1e-12`, they do. However, they do not in the test case. All of them precisely align the primary axis, but then the rotation about that axis is inconsistent. See the below plot, looking down the aligned primary axes onto the projected secondary plane. If I calculate the ratios of the angular error for the third and second vectors in this plane, I get `[3.1876, 1.6074, 2.0264]`, whereas I would expect a perfect ratio of 2 from the weights.
![image](https://github.com/scipy/scipy/assets/14363975/22bd30ef-2ced-4138-a81f-e3fc0054cb3a)

So I'm a good bit stuck. I'm having trouble finding a canonical case in the literature that I can replicate. Do you have any ideas? If not then for right now I would propose allowing only the N=2 case for an infinite weight, until we can figure out the right behavior for N>2.

scottshambaugh commented:
Some more work on this today, and I've started to get results that line up with the align_vectors implementation. The differences above have to do with the cost function / weights being calculated on the full vectors vs the vectors as projected into the plane pierced by the primary axis. Some more work to do on my end before this is ready.

scottshambaugh commented:
Hi @nmayorov I'm happy to report that I got the N>2 case working as expected - one infinite weight with this align-constrain process gives the same result as the limit of the baseline process as one of the weights approaches infinity. I think this is ready for review now.

nmayorov commented:
> Hi @nmayorov I'm happy to report that I got the N>2 case working as expected - one infinite weight with this align-constrain process gives the same result as the limit of the baseline process as one of the weights approaches infinity. I think this is ready for review now.
> 
> There was some git weirdness when I was rebasing onto main, these commits can all be squashed during the merge.

Sounds good, reviewing now.

scottshambaugh commented on scipy/spatial/transform/_rotation.pyx:
```diff
@@ -2766,36 +2805,78 @@ cdef class Rotation:
+        vector against each of the secondary non-infinite weight vectors.
+        These rotations are then weighted averaged together to get the final
+        result. The result via this process is the same as the Kabsch
+        algorithm in the limit as the corresponding weight approaches infinity
```
double "in the limit", clean this up

scottshambaugh commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1315,20 +1335,128 @@ def test_align_vectors_invalid_input():
+    mats = []
+    for i in range(6):
+        mats.append(Rotation.random(n, random_state=10 + i).as_matrix())
+
```
TODO: update to non-unit length

scottshambaugh commented on scipy/spatial/transform/_rotation.pyx:
```diff
@@ -2806,54 +2886,257 @@ cdef class Rotation:
+            # the infinite weight vector, so we force that to avoid issues when
+            # multiplying by infinity
+            weights_non_inf = weights.copy()
+            weights_non_inf[weight_is_inf] = 0
```
There is an issue here if the corresponding a and b vectors are not of equal length. In that instance the error will not be 0, and this will not line up with the baseline algorithm in the limit. I don't have a good idea of how to avoid this, without changing the behavior everywhere to calculate rssd from the normalized vectors.

nmayorov commented:
To be honest, I think your method of combining multiple "two vector rotations" using  `mean` with special weighing is a bit superfluous and not very transparent mathematically. As I pointed out before, the secondary alignment comes down to solving a 2D alignment problem. 

I've written a working sketch of such algorithm, so you can evaluate (and hopefully appreciate) its flow:

```Python
import numpy as np
from scipy.spatial.transform import Rotation


def row_wise_dot(a, b):
    return np.sum(a * b, axis=1)


def align_vectors_inf(a, b, weights):
    a = np.asarray(a, dtype=float)
    b = np.asarray(b, dtype=float)
    weights = np.array(weights, copy=True, dtype=float)
    if len(weights) == 1:  # For conding purposes to save some ifs later
        weights[0] = np.inf

    is_inf = np.isinf(weights)
    n_inf = np.sum(is_inf)
    if n_inf > 1:
        raise ValueError("Only a single weight can be infinite")

    if n_inf == 0:
        return Rotation.align_vectors(a, b, weights)

    a_pri = a[is_inf][0]
    b_pri = b[is_inf][0]
    a_pri /= np.linalg.norm(a_pri)
    b_pri /= np.linalg.norm(b_pri)
    
    # This is the block to find the minimal rotation between two vectors.
    cross = np.cross(b_pri, a_pri)
    theta = np.arctan2(np.linalg.norm(cross), np.dot(a_pri, b_pri))
    if theta < 1e-3:  # small angle Taylor series approximation
        theta2 = theta * theta
        r = cross * (1 + theta2 / 6 + theta2 * theta2 * 7 / 360)
    else:
        r = cross * theta / np.sin(theta)
    R_pri = Rotation.from_rotvec(r)
    
    # Even with a single pair of vectors (~is_inf is empty) the code below is correct and phi will be zero.
    a = a[~is_inf]
    b = b[~is_inf]
    c = R_pri.apply(b)
    weights = weights[~is_inf]
    phi = np.arctan2(np.sum(weights * row_wise_dot(np.cross(c, a), a_pri)),
                     np.sum(weights * (
                             row_wise_dot(c, a) - row_wise_dot(c, a_pri) * row_wise_dot(a, a_pri))))
    R_sec = Rotation.from_rotvec(phi * a_pri)

    return R_sec * R_pri
```

This code passes your tests which compare the results with the standard algorithm with weight of 1e12. The special weighing you've come up with occurs here naturally when computing projections of secondary vectors (where `phi` is computed).

As you can see, the constrained and minimal (for single pair of vectors) alignment comes down to finding primary rotation and then the secondary rotation (which will be trivial when there are no secondary vectors). I'm quite sure it will work significantly faster and the flow overall looks cleaner to me.

scottshambaugh commented on scipy/spatial/transform/_rotation.pyx:
```diff
@@ -2806,54 +2833,213 @@ cdef class Rotation:
+            # the infinite weight vector, so we force that to avoid issues when
+            # multiplying by infinity
+            weights_inf_zero = weights.copy()
+            weights_inf_zero[weight_is_inf] = 0
```
There is an issue here if the corresponding a and b vectors are not of equal length. In that instance the error will not be 0, and this will not line up with the baseline algorithm. I don't have a good idea of how to avoid this, without changing the behavior everywhere to calculate rssd from the normalized vectors.

nmayorov replied:
I agree with you. I think the correct view is not to consider the pair with the infinite weight as belonging to the cost function. It works as a constraint.

I'd say the only edge case is a single pair of vectors which can be viewed (sort of) either way. I think any solution will be fine:

1. Set `rssd` to zero and mention it in the docstring.
2. Compute it normally, but then forbid an infinite weight when only one pair of vectors is aligned (and perhaps adjust the code with artificial infinite weight).

Consider both options.

scottshambaugh replied:
I think setting `rssd` to 0 makes the most sense (and is the current behavior). For case 2, we still run into the same problem if the primary vectors are not of the same length, even if there are secondary vectors as well. For case 3, I think it would be rather confusing since I feel like the intuitive view is to see the weights as relative rather than absolute. So a weight of 1 would be in contrast to a weight of 0 and still be relatively infinite.

Added some explanation to the docstring that we force this component of rssd to be 0.

scottshambaugh replied:
That makes sense to me, and was not awkward to implement - only needed one if statement.

scottshambaugh commented:
@nmayorov thank you! That approach is definitely much more elegant & faster, and I can confirm that it gives the same correct results. It also allows removing the `atol` parameter, which is excellent. I updated one of my comments above to save the old approach for posterity, but have updated the MR to use your code. I must admit that I still don't fully understand the meaning of the vector calculations when finding the secondary rotation angle phi, so will have to digest that.

scottshambaugh commented:
I sketched it out and understand the calculation for the secondary rotation angle. Added some comments in the code to explain it, and I think this is quite a good state now.

nmayorov commented on scipy/spatial/transform/_rotation.pyx:
```diff
@@ -2806,54 +2833,237 @@ cdef class Rotation:
+                # Calculate vector components for the angle calculation. The
+                # angle phi to rotate a single 2D vector C to align to 2D
+                # vector A in the xy plane can be found with the equation
+                # phi = atan2(cross(C, A), dot(C, A)).
```
Yes, it's true. And if you have many vectors the same formula can be applied, but with sums over all vectors (possibly weighted). It can be derived rigorously (in 2D) by finding a stationary point of a function of one argument (the angle phi).

The trick in 3D is to consider projections of vectors into the plane. Let's denote the normal unit vector as `u` (in our case is `a_pri`). Let `Cp` and `Ap` be the orthogonal projections. Because they still live in 3D we have to project the cross product onto `u`, by using `(Cp x Ap) . u`. For dot product we don't have to change anything and simply `Cp . Ap`.

Now is we substitute `Ap = A - (A . u) u` and `Cp = C - (C . u) u` (remove the component along the `u` to get the orthogonal projection) we will get the formulas you've used here. They are just an alternative form to get the angle phi without explicitly computing `Cp` and `Ap` (by the way I think the code where `Cp` and `Ap` are computed would be also fine).

nmayorov commented on scipy/spatial/transform/_rotation.pyx:
```diff
@@ -2806,54 +2833,237 @@ cdef class Rotation:
+               [0., 1., 2.]])
+        """
+        # Check input vectors
+        a = np.array(a, dtype=float)
```
I think here we should call `atleast_2d`  right here.

scottshambaugh replied:
This needed a little rework to make the input checks report out the actual user input, but the logic makes sense now.

nmayorov commented on scipy/spatial/transform/_rotation.pyx:
```diff
@@ -2806,54 +2833,237 @@ cdef class Rotation:
-
+            weights = np.array(weights, dtype=float)
+
+        if weights.ndim != 1:
```
Well, I think it's fine to keep these checks inside `else` (as was before). It seems correct and logical as well, better not to create extra diff here.

scottshambaugh replied:
Makes sense to me, fixed.

nmayorov commented on scipy/spatial/transform/_rotation.pyx:
```diff
@@ -2806,54 +2833,237 @@ cdef class Rotation:
+            if return_sensitivity:
+                raise ValueError("Cannot return sensitivity matrix with an "
+                                 "infinite weight or one vector pair")
+            a_pri = a[weight_is_inf]
```
Why not take `[0]` right here?

Also, you of course may adjust the names as you see fit, `a_pri` and `b_pri` seems fine though.

scottshambaugh replied:
einsum down below needs this to be a 2d input to work, and I feel like it's cleaner as-is rather than converting down below

nmayorov commented on scipy/spatial/transform/_rotation.pyx:
```diff
@@ -2806,54 +2833,237 @@ cdef class Rotation:
+                                 "infinite weight or one vector pair")
+            a_pri = a[weight_is_inf]
+            b_pri = b[weight_is_inf]
+            a_pri /= _norm3(a_pri[0])
```
This is a somewhat annoying subject, but what if `a_pri` or `b_pri` will be exactly zero? Perhaps check for exact zero and raise an error?

scottshambaugh replied:
Yeah that would be annoying. Added a check. I believe the rest of the algorithm should work fine if any of the secondary vectors was zero.

nmayorov commented on scipy/spatial/transform/_rotation.pyx:
```diff
@@ -2806,54 +2833,237 @@ cdef class Rotation:
+            R_pri = cls.from_rotvec(r)
+
+            if N == 1:
+                # While the else clause below will work for a single pair of
```
I think this shortcut is totally intuitive and perhaps it would be strange for someone if we do otherwise. 

So maybe this comment is not necessary.

scottshambaugh replied:
Cut this down

nmayorov commented on scipy/spatial/transform/_rotation.pyx:
```diff
@@ -2806,54 +2833,237 @@ cdef class Rotation:
+                # product of X and Y.
+                S = np.einsum('ij,ij->i',
+                              np.cross(c_non_inf, a_non_inf), a_pri)
+                T = (np.einsum('ij,ij->i', c_non_inf, a_non_inf)
```
Any particular reason to use name `T` here?

scottshambaugh replied:
Nope, this was arbitrary. I changed the name here a bit to hopefully make it easier to follow

nmayorov commented on scipy/spatial/transform/_rotation.pyx:
```diff
@@ -2806,54 +2833,237 @@ cdef class Rotation:
+                # The weighted sum of the components essentially gives the
+                # average of all the rotation angles for each secondary
+                # vector pair. This average is the rotation angle about the
+                # primary axis that best aligns all the secondary vectors
```
Well, if possible try to combine this and the previous comment to make it more clear.

I think that just saying that we solve a 2D alignment problem is mostly sufficient.

By the way, this approach of finding the mean direction (angle) is used in circular statistics, where the average sine and cosine is found first.

scottshambaugh replied:
Consolidated these comments a bit. I do think having some exposition here is useful for whoever else reads this code, since I think vector math can be pretty inscrutable if presented without context.

nmayorov reviewed (commented):
I think it is mostly finished. The updated doctrings and examples are on point.

Some minor things to consider.

nmayorov reviewed (commented):
I plan to check the tests soon.

scottshambaugh commented on scipy/spatial/transform/_rotation.pyx:
```diff
@@ -2813,47 +2842,230 @@ cdef class Rotation:
+        # Check input vectors
+        a_original = np.array(a, dtype=float)
+        b_original = np.array(b, dtype=float)
+        a = np.atleast_2d(a_original)
```
I like the current version better since it is slightly more permissive on inputs, so will leave as-is unless you feel strongly.

scottshambaugh commented on scipy/spatial/transform/_rotation.pyx:
```diff
@@ -2813,47 +2842,230 @@ cdef class Rotation:
+                # 3D rotation restricted to the 2D plane normal to a_pri, where
+                # the secondary vectors are projected into that plane. The
+                # weighted sum of the components gives the average of all the
+                # angles.
```
Updated the wording. Yeah I agree that a circular average is not intuitive to many people.

nmayorov commented:
@scottshambaugh 

Happy to merge this pull request. I've updated the [release notes](https://github.com/scipy/scipy/wiki/*Release-Note-Entries-for-SciPy-1.12.0-(asterisk-makes-this-appear-at-the-top)). Check if you agree with the description and whether you think the new minimal rotation should be mentioned in "Backwards incompatible changes" (I haven't done it).

Ping me if you'll have other pull requests to consider.

scottshambaugh commented:
Thank you for the collaboration on this, I think the result is much better for it! I have no other pull requests at this time.

The release notes look good. I would not consider this backwards incompatible behavior, since if you tried to align shape (3,) vectors previously you would get an error, and if you tried to align shape (1, 3) vectors you get the below warning. So the previous behavior had no uniqueness guarantees.
```
UserWarning: Optimal rotation is not uniquely or poorly defined for the given sets of vectors.
```

nmayorov commented:
>The release notes look good. I would not consider this backwards incompatible behavior, since if you tried to align shape (3,) vectors previously you would get an error, and if you tried to align shape (1, 3) vectors you get the below warning. So the previous behavior had no uniqueness guarantees.

These are good points, thank you!
